### PR TITLE
New feature restricting what passwords can be updated to based on history

### PR DESCRIPTION
**Pull request by ryancutter**
Current UAA behavior states users cannot update passwords to the current password.  So this is essentially a password history of "1".  This PR seeks to make this more flexible, allowing for a history of "0" or any number.

### CODE REVIEW

**Comment by ryancutter** [comment]
Not ready for review, need to add tests.  Created this PR for collaboration purposes.

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/PasswordChangeEndpoint.java [hunk 1]
```diff
@@ -12,10 +12,10 @@
  *******************************************************************************/
 package org.cloudfoundry.identity.uaa.account;
 
+import static org.springframework.http.HttpStatus.UNPROCESSABLE_ENTITY;
+
```
Comment (ryancutter) [inline server/src/main/java/org/cloudfoundry/identity/uaa/account/PasswordChangeEndpoint.java]: IJ settings messed with order, we can fix this up.

File: server/src/main/java/org/cloudfoundry/identity/uaa/account/PasswordChangeEndpoint.java [hunk 2]
```diff
@@ -87,11 +96,8 @@ public ActionResult changePassword(@PathVariable String userId, @RequestBody Pas
         String newPassword = change.getPassword();
 
         throwIfPasswordChangeNotPermitted(userId, oldPassword, zoneId);
-        if (dao.checkPasswordMatches(userId, newPassword, zoneId)) {
-            throw new InvalidPasswordException("Your new password cannot be the same as the old password.", UNPROCESSABLE_ENTITY);
-        }
```
Comment (ryancutter) [inline server/src/main/java/org/cloudfoundry/identity/uaa/account/PasswordChangeEndpoint.java:92]: This is the old behavior, of course.  Now added to `changePasswordWithHistoryCheck()`.  Figured the history check should not be executed until the original password is authenticated as to not leak comprising error messages unnecessarily.

Similar type changes in the next 2 classes as well.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 3]
```diff
@@ -281,17 +293,38 @@ public void changePassword(final String id, String oldPassword, final String new
...
+
+        changePasswordAndWriteHistory(id, newPassword, zoneId);
+    }
+
+    @Override
+    public void changePasswordWithHistoryCheck(final String id, final String oldPassword, final String newPassword, final String zoneId,
+            final int historyLength) throws ScimResourceNotFoundException {
+        if (oldPassword != null && !checkPasswordMatches(id, oldPassword, zoneId)) {
+            throw new BadCredentialsException("Old password is incorrect");
+        }
```
Comment (ryancutter) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java:304]: So this above block is repeated in `checkPassword()`, suggestions welcome.  Note this is to get the following check sequenced correctly.

File: server/src/main/resources/spring/login-ui.xml [hunk 4]
```diff
@@ -554,6 +554,7 @@
     <bean id="changePasswordService" class="org.cloudfoundry.identity.uaa.account.UaaChangePasswordService">
         <constructor-arg ref="scimUserProvisioning"/>
         <constructor-arg ref="uaaPasswordValidator"/>
+        <constructor-arg value="${password.historyRestriction:1}"/>
```
Comment (ryancutter) [inline server/src/main/resources/spring/login-ui.xml:557]: A history of 1 mimics existing UAA behavior in which you cannot update the password to the same value.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 5]
```diff
@@ -521,4 +580,17 @@ protected void validateOrderBy(String orderBy) throws IllegalArgumentException {
...
+        // add new password to history table
+        jdbcTemplate.update(ADD_PASSWORD_HISTORY_SQL, new PreparedStatementSetter() {
+            @Override
+            public void setValues(PreparedStatement ps) throws SQLException {
+                ps.setString(1, id);
+                ps.setString(2, zoneId);
+                ps.setString(3, encNewPassword);
+                ps.setTimestamp(4, timestamp);
+            }
+        });
```
Comment (ryancutter) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java:593]: So we're just adding encoded passwords to this history table.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 6]
```diff
@@ -89,6 +91,10 @@ public Log getLogger() {
 
     public static final String CHANGE_PASSWORD_SQL = "update users set lastModified=?, password=?, passwd_lastmodified=? where id=? and identity_zone_id=?";
 
+    public static final String ADD_PASSWORD_HISTORY_SQL = "insert into password_history (user_id, identity_zone_id, password, changed) values (?,?,?,?)";
+
+    public static final String GET_PASSWORD_HISTORY_SQL = "select password from password_history where user_id=? and identity_zone_id=? order by changed desc limit ?";
+
```
Comment (ryancutter) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java:97]: These are the 2 new queries - adding the password to the history table and fetching passwords for a particular user in temporal order.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ForcedPasswordChangeIT.java [hunk 7]
```diff
@@ -125,7 +125,7 @@ public void testHandleForceChangingPasswordWithNewPasswordSameAsOld() throws Exc
         webDriver.findElement(By.xpath("//input[@value='Create new password']")).click();
         assertEquals(baseUrl+"/force_password_change", webDriver.getCurrentUrl());
         assertThat(webDriver.findElement(By.cssSelector(".error-message")).getText(),
-            containsString("Your new password cannot be the same as the old password."));
+            containsString("Your new password cannot be the same as one in your recent password history."));
```
Comment (ryancutter) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/ForcedPasswordChangeIT.java:128]: Not really sure how to run the ITs.  Can someone send a link with instructions?  Thanks.

File: server/src/main/resources/org/cloudfoundry/identity/uaa/db/hsqldb/V4_22_0__Add_Password_History.sql [hunk 8]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE password_history (
```
Comment (ryancutter) [inline server/src/main/resources/org/cloudfoundry/identity/uaa/db/hsqldb/V4_22_0__Add_Password_History.sql:1]: Oh just noticed these versions should perhaps be `V4_24_0` to align with the release?

**Comment by shamus** [comment]
Closing due to inactivity.

**Pull request closed without merging**